test/demo/disconnect.inttest.ts

```diff
@@ -0,0 +1,174 @@
+import {setupTeardown, quitApps, TestAppData} from './utils/common';
+import {testManagerIdentity, testAppInDirectory1, testAppWithPreregisteredListeners1, testAppNotInDirectory1, testAppNotFdc3} from './constants';
+import {RemoteChannel} from './utils/RemoteChannel';
+import {delay} from './utils/delay';
+
+setupTeardown();
+
+export type ProviderWindow = Window & {
+    model: Model;
+    intentHandler: IntentHandler;
+    channelHandler: ChannelHandler;
+    eventHandler: EventHandler;
+}
+
+const ofBrowser = new OFPuppeteerBrowser();
+const TEST_INTENT = 'TestIntent';
+let redChannel: RemoteChannel;
+let blueChannel: RemoteChannel;
+
+type TestParam = [
+    string,
+    TestAppData,
+    (app: TestAppData) => Promise<void>,
+    (app: TestAppData) => Promise<void>
+];
+
+type TestCatagoryParam = [string, TestParam[]];
+
+async function openDirectoryApp(app: TestAppData) {
+    await fdc3Remote.open(testManagerIdentity, app.name);
+}
+
+async function openNonDirectoryApp() {
+    await fin.Application.startFromManifest(testAppNotInDirectory1.manifestUrl);
+}
+
+const directoryApps: TestParam[] = [
+    ['closed', testAppInDirectory1, openDirectoryApp, async (app) => {
+        await quitApps(app);
+    }],
+    ['navigated away', testAppInDirectory1, openDirectoryApp, async (app) => {
+        await navigateTo(app, 'about:blank');
+    }],
+    ['reloaded', testAppInDirectory1, openDirectoryApp, async (app) => {
+        await reload(app);
+    }],
+    ['closed with preregistered listeners', testAppWithPreregisteredListeners1, openDirectoryApp, async (app) => {
+        await quitApps(app);
+    }],
+    ['navigated away with preregisted listeners', testAppWithPreregisteredListeners1, openDirectoryApp, async (app) => {
+        await navigateTo(app, 'about:blank');
+    }],
+    ['reloaded with preregistered listeners', testAppWithPreregisteredListeners1, openDirectoryApp, async (app) => {
+        await reload(app);
+    }]
+];
+
+const nonDirectoryApps: TestParam[] = [
+    ['closed', testAppNotInDirectory1, openNonDirectoryApp, async (app) => {
+        await quitApps(app);
+    }],
+    ['navigated away', testAppNotInDirectory1, openNonDirectoryApp, async (app) => {
+        await navigateTo(app, 'about:blank');
+    }],
+    ['reloaded', testAppNotInDirectory1, openNonDirectoryApp, async (app) => {
+        await reload(app);
+    }]
+];
+
+const testCatagories: TestCatagoryParam[] = [['Directory', directoryApps], ['Non Directory', nonDirectoryApps]];
+
+describe('Disconnecting windows', () => {
+    // Directory Apps or NonDirectory Apps
+    describe.each(testCatagories)('%s Apps', (catagoryTitle: string, tests: TestParam[]) => {
+        // E.g. When an app is closed.
+        describe.each(tests)('When an app is %s', (
+            testTitle: string,
+            application: TestAppData,
+            openMethod: (app: TestAppData) => Promise<void>,
+            disconnectMethod: (app: TestAppData) => Promise<void>
```

Comment:
Minor terminology quibble - method says to me "function on a class instance". Consider just going with `openFunction`/`disconnectFunction`?

Reply:
Switched it